src/arc_dst.rs

```diff
@@ -0,0 +1,476 @@
+
+impl<T: Copy> RefCount<[T]> {
+    fn from_slice(slice: &[T]) -> Box<RefCount<[T]>> {
+        use std::alloc::Layout;
+        let layout = Layout::new::<RefCount<()>>()
+            .extend(Layout::array::<T>(slice.len()).unwrap())
+            .unwrap()
+            .0
+            .pad_to_align();
+        // SAFETY: the layout is not zero-sized, it at least has an `AtomicUsize`.
+        let ptr = unsafe { std::alloc::alloc(layout) };
+        // imitate the `std::Arc::new_zeroed_slice` and `std::Arc::try_allocate_for_layout`.
+        let ptr =
+            std::ptr::slice_from_raw_parts_mut(ptr as *mut T, slice.len()) as *mut RefCount<[T]>;
+
+        unsafe {
+            // SAFETY: the ptr is allocated by the global allocator with proper layout, as per the
+            // [Memory Layout](https://doc.rust-lang.org/stable/std/boxed/index.html#memory-layout) section
+            // of [`std::boxed`].
+            let mut this = Box::from_raw(ptr);
+
+            std::ptr::write(&mut this.count, AtomicUsize::new(1));
+
+            // SAFETY: valid for reads, writes, aligned and not overlapped.
+            // and T is Copy, so don't worry about drop.
+            let dst = &mut this.data as *mut [T] as *mut T;
+            std::ptr::copy_nonoverlapping(slice.as_ptr(), dst, slice.len());
+            this
+        }
+    }
+}
+
+impl RefCount<str> {
+    fn from_str(s: &str) -> Box<RefCount<str>> {
+        let bytes = s.as_bytes();
+        let boxed_refcount = RefCount::<[u8]>::from_slice(bytes);
+        debug_assert_eq!(s.len(), boxed_refcount.data.len());
+
+        // SAFETY: str and [u8] have the same memory layout.
+        unsafe { Box::from_raw(Box::into_raw(boxed_refcount) as *mut RefCount<str>) }
+    }
+}
+
+impl<T: ?Sized + Eq + Hash + Send + Sync + 'static> ArcIntern<T> {
+    /// make new [`ArcIntern`] with copyable initial value, like `&str` or `&[u8]`.
+    fn new_with_copyable_init_val<I, NewFn>(val: &I, new_fn: NewFn) -> ArcIntern<T>
+    where
+        I: ?Sized + Hash + std::cmp::Eq,
+        BoxRefCount<T>: Borrow<I>,
+        NewFn: Fn(&I) -> Box<RefCount<T>>,
+    {
+        // cache the converted BoxRefCount to avoid copy. This only takes an usize.
+        let mut converted = None;
+        loop {
+            let m = Self::get_container();
+
+            if let Some(b) = m.get_mut(val) {
+                let b = b.key();
+                // First increment the count.  We are holding the write mutex here.
+                // Has to be the write mutex to avoid a race
+                let oldval = b.0.count.fetch_add(1, Ordering::SeqCst);
+                if oldval != 0 {
+                    // we can only use this value if the value is not about to be freed
+                    return ArcIntern {
+                        pointer: std::ptr::NonNull::from(b.0.borrow()),
+                    };
+                } else {
+                    // we have encountered a race condition here.
```

Comment:
I wish there were a way to structure the code without duplicating this logic.  I don't see that way, but if you do see a way to make one function do this tricky race handling that would be awesome.  My concern is that if a bug is found it might not get fixed in both places.  Or if any change is made.

Reply:
Yeah I agree, it would be great if there's a way to avoid the duplicate. But currently I dont see how that's possible